core/store/src/trie/mod.rs

```diff
@@ -800,6 +810,87 @@ mod tests {
 
     use super::*;
 
+    fn verify_proof(
```

Comment:
Perhaps we should move this alongside with the test veryfying proof to integration-tests/src/tests/runtime/state_viewer.rs with the other view state tests.  This is not used by the node for anything.

Also, verifying the proof is now a bit more complex since we’re verifying all the items, not just the one at prefix.  To verify proof we need: state root hash, the proof, prefix and all the returned values.  Verification would need to first hash all the nodes and build a hash→node mapping and from that it can start verifying values.

```rust
struct ProofVerifier {
    nodes: HashMap<CryptoHash, TrieNodeWithSize>,
}

impl ProofVerifier {
    fn new(proof: Vec<Arc<[u8]>>) -> Self {
        let nodes = proof.into_iter().map(|bytes| {
            let hash = CryptoHash::hash_bytes(&bytes);
            let node = RawTrieNodeWithSize::decode(&bytes);
            (hash, node)
        }).collect::<HashMap<_, _>>();
        Self { nodes }
    }

    fn verify(
        state_root: &StateRoot,
        account_id: &AccountId,
        key: &[u8],
        expected: Option<&[u8]>,
    ) -> bool;
}

verify would then traverse the nodes in the hash map to retrieve the value at given key and compare it to what is expected.  When doing the verification the actual trie key would need to be properly constructed as per get_raw_prefix_for_contract_data.

Reply:
The problem with this is that some structs aren't public (i.e. `RawTrieNodeWithSize`)

Reply:
would you be ok duplicating the definitions?

Reply:
Just make whatever needs to be public public. :)